### PR TITLE
List View: Use heading content for button label text if available

### PR DESCRIPTION
**Pull request by andrewserong**
## What?

Explore conditionally using the Heading block's content as the block button label in the List View, falling back to the block's title if no content is available.

## Why?

As discussed in https://github.com/WordPress/gutenberg/issues/33583, this could help improve navigating around the List View, and make it feel a little more like working with a document outline.

## How?

* Update the `BlockTitle` component and `useBlockDisplayTitle` hook to support passing in a `context` string that will then be passed along to `getBlockLabel`.
* In the Heading block, check for a `list-view` context string and then conditionally return the block's content as the label.
* In the List View, pass in `list-view` as the context to `BlockTitle` for the block button.

## Testing Instructions

1. Insert a heading block into a post/page
2. When the heading is empty it should say "Heading" in the list view
3. When the heading has any content in it, it should use that (truncated) in the list view
4. In the contextual menu for selecting a block within the editor canvas, it should still say "Remove Heading" and _not_ use the block's content as a label
5. Other blocks should display in the list view as before

## Screenshots or screencast 

In the below screengrab, the empty Heading block defaults to using the block name as the label. As the content is updated, the label updates to use the block's content.

| Before | After |
| --- | --- |
| image | image |

### CODE REVIEW

File: packages/block-editor/src/components/list-view/block-select-button.js
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/block-select-button.js]: It might be a bit much including the `paragraph` here, so we might want to pare this back to just the `heading` block for now 🤔

**Comment by talldan** [comment]
There's an existing system for doing this that should make the PR a bit smaller.

You can see how the navigation link block achieves it here - https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/navigation-link/index.js#L25.

File: packages/block-editor/src/components/list-view/block-select-button.js
Comment (mtias) [reply packages/block-editor/src/components/list-view/block-select-button.js]: Let's start with heading, I think paragraph might be worth treating differently:

image

**Comment by mtias** [comment]
@talldan that reminds me, we should probably _unexperimentalize_ that API :)

File: packages/block-editor/src/components/list-view/block-select-button.js
Comment (andrewserong) [reply packages/block-editor/src/components/list-view/block-select-button.js]: I've updated so that it only affects the Heading block for now. I'm happy to explore that Paragraph idea in a follow-up if no-one beats me to it, looks like it could be a good way to handle it!

**Comment by andrewserong** [comment]
Thanks @talldan, using that function's a great idea! I've updated this PR — I went a little further and added a `context` param to the `BlockTitle` and `useBlockDisplayTitle` so that we can conditionally use the block content only in the list view. Let me know if you don't think this is the right way to handle it, but I thought that could be a neat way to ensure that the block name is still used in other places (such as the Remove button in the block settings dropdown in the editor canvas).

**Comment by talldan** [comment]
> I went a little further and added a context param to the BlockTitle and useBlockDisplayTitle so that we can conditionally use the block content only in the list view. 

I wonder if that'll be needed. I think the only other place it'll display is the breadcrumb at the bottom of the screen. I think other blocks that use `__experimentalLabel` show their label in the breadcrumb.

It might be good to get input from a designer about this part.

**Comment by andrewserong** [comment]
> I wonder if that'll be needed

Good question, it appears to be used for both the breadcrumb at the bottom of the screen and the Remove button in the dropdown (the settings dropdown uses `useBlockDisplayTitle` [here](https://github.com/WordPress/gutenberg/blob/6f3bb8c44e9443e91c696f71a84edb9454bbe910/packages/block-editor/src/components/block-settings-menu/block-settings-dropdown.js#L127)). With some longer titles it felt a bit odd to me using the content as the label, whereas in the List view it feels a bit clearer to me since it has the Heading icon. Here's how it looks if we don't have the check for `list-view` context:

image

I'll add a "Needs Design feedback" label.

TL;DR for designers: should we use Heading content as the label just in the list view, or is it also okay in the breadcrumb at the bottom of the screen, and in the remove button in the settings menu? (I'm happy to implement it either way 🙂)

**Comment by talldan** [comment]
> Good question, it appears to be used for both the breadcrumb at the bottom of the screen and the Remove button in the dropdown (the settings dropdown uses useBlockDisplayTitle [here](https://github.com/WordPress/gutenberg/blob/6f3bb8c44e9443e91c696f71a84edb9454bbe910/packages/block-editor/src/components/block-settings-menu/block-settings-dropdown.js#L127)).

Hmm, yeah, it doesn't look great in the remove button. There may need to be some more nuance in how this is shown. I think the distinction is in whether the title shows content (like it will for Heading and does for Navigation Link), or whether it represents the semantic name of a block (like it does for variations and the names of template parts / reusable blocks).

**Comment by andrewserong** [comment]
> I think the distinction is in whether the title shows content

Ah, so we could potentially use a context value of `content` instead of `list-view`?

**Comment by talldan** [comment]
I think the problem with really specific terms like 'list-view' is that this could grow into a really big list of contexts as different types of UI are introduced, or ideas around this are iterated on.

All in all, I'm not sure this context system is the right solution in the first place. Maybe it should be something like this pseudocode:
```js
getLabels( blockType, attributes ) {
    return {
        // used where there's a short amount of space for the block (e.g. remove block button)
        title: `${ blockType.title } (${ attributes.level })`,
        // used where there's more space (list view, breadcrumbs)
        label: attributes.content,
        // used for aria labels
        accessibility: `${ attributes.level } - ${ attributes.content }`,
    };
}
```

These outstanding issues are one of the reasons why it hasn't been stabilised yet

**Comment by andrewserong** [comment]
> These outstanding issues are one of the reasons why it hasn't been stabilised yet

Ah, gotcha. And then it's also potentially tricky if we want different behaviour for different blocks (like in the screenshot of how the Paragraph block might work: https://github.com/WordPress/gutenberg/pull/41855#discussion_r903556163), which appears to be a strength of the existing API 🤔. If we wind up wanting to proceed with the label in the list view being different from the remove button or breadcrumbs, but it's too fiddly to land on a change to that API, I can always roll this PR back to the initial commit.

Alternately, if it's acceptable that the Remove button also uses the content, then there's a clearer path to a smaller changeset for this PR (just change the `__experimentalLabel` for the heading block for all visual contexts, and don't mess with the BlockTitle component 🙂).

Thanks again for thinking through the implications, and I'm happy to keep hacking away tomorrow!

**Comment by mtias** [comment]
My inclination would be to use only on list-view. Not quite sure about breadcrumbs, but I lean towards no.

**Comment by andrewserong** [comment]
Thanks for the feedback! I ran out of time to progress this one today, but I'll revisit next week. I suspect we'll need a little more logic to make sure that long titles truncate comfortably for headings with a custom anchor `id` in the button label, too, so as a first step, it might be worth us keeping the logic encapsulated to the List View button component for now (with a hard-coded list of blocks we're dealing with similar to my first commit), also so that we can side-step the discussion about the shape of the label API.

That said, do let me know if anyone feels strongly about how it should be implemented, though, and I'll incorporate the feedback when I revisit this next week. Thanks, all! 🙇

**Comment by andrewserong** [comment]
I've updated this PR to revert back to the approach of using some hard-coded logic in the `block-select-button` component, so that we can side-step the problem of the label API for the moment.

I took a look at how to get the truncation to play nicely with the `id` attribute, and it looks like if we add an `HStack` wrapper over the title and anchor tag, we can get the truncation to factor in the existence of that pill. As an example, here's a screenshot of the current state of the PR:

image

In the above screenshot, from the highlighted Heading:

* A long heading that gets truncated to still allow room for the HTML anchor.
* A long heading that gets truncated.
* A long heading that gets truncated and is locked (shows how it sits with a lock icon)
* A reusable block that gets truncated (no change here, just demonstrating)

In each of the above examples, space is effectively "reserved" for the lock icon via the max width settings. If the lock is removed on that particular heading, then the whole List View returns to the default width of `350px`.

To achieve the above, I've added a `max-width` of `225px` to the wrapper containing the title + anchor id, which appears to be the maximum width before the long titles start to extend over the minimum width of the List View container (`350px`) — there might be a better width to go with (either a little less or a little more), but at least in my testing today, it seemed to be a kind of sweet spot between too narrow, or letting long headings stretch out the list view to be a bit too wide. Very happy for feedback or ideas on an ideal max width.

The PR probably still needs a little bit of tidying, but happy to try implementing any other ideas. I think the additional `HStack` with the max-width should help us with the Paragraph block follow-up, too, where we'll try out having grey truncated content text next to the block title (as in this screenshot: https://github.com/WordPress/gutenberg/pull/41855#discussion_r903556163).

(Note: I think the current change possibly breaks the e2e tests, so I'll look at fixing that up, too)

**Comment by talldan** [comment]
> I've updated this PR to revert back to the approach of using some hard-coded logic in the block-select-button component, so that we can side-step the problem of the label API for the moment.

Sorry to potentially make you u-turn again, but I do think it'd be better to use the existing label API. The `context === 'list-view'` approach you outlined is a reasonable trade-off right now if there are no other quick wins. At least this means that all the blocks that have a label are implemented similarly, and there's then a basis from which to improve the whole API.

I don't think defining the block names implicitly in list view is the way to go - it wouldn't be something that could easily develop as an API because it limits the feature to core blocks.

**Comment by andrewserong** [comment]
> Sorry to potentially make you u-turn again, but I do think it'd be better to use the existing label API. The context === 'list-view' approach you outlined is a reasonable trade-off right now if there are no other quick wins.

Not a bother at all @talldan, it's easy to revert! I'll rework this PR to go with the `list-view` context approach later on today, and it was still useful figuring out the `max-width` stuff yesterday, so no wasted effort 🙂

**Comment by andrewserong** [comment]
Just a note for myself, I think the e2e tests are failing because of the additional `div` so the title span isn't a direct child of the `a` tag in the list view. I'll probably need to update the logic in: https://github.com/WordPress/gutenberg/blob/trunk/packages/e2e-test-utils/src/get-list-view-blocks.js#L7, and maybe try to make it a little more flexible. I'll look at that tomorrow.

File: packages/e2e-test-utils/src/get-list-view-blocks.js
Comment (andrewserong) [inline packages/e2e-test-utils/src/get-list-view-blocks.js]: This change means that it looks for `span` as any child of the `a` tag instead of a direct child. This allows us to adjust the level of nesting between the `a` tag and the `blockLabel` text without breaking tests that depend on this function (🤞)

Edit: I updated the logic slightly again in: https://github.com/WordPress/gutenberg/pull/41855/commits/7344bed5fbc6c469264fea9bd9e7941933a81b00

File: packages/block-editor/src/components/list-view/block-select-button.js
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/block-select-button.js]: I've switched this to use `Truncate` instead of `BlockTitle` — in this context, due to the variable width of the available space, we want to use the `auto` ellipsizeMode instead of truncating to a particular number of characters, so this way gives us a bit more flexibility.

Comment (andrewserong) [reply packages/block-editor/src/components/list-view/block-select-button.js:85]: I see I've missed a couple of tests that also use XPaths expecting a particular sibling of the label. E.g.

* https://github.com/WordPress/gutenberg/blob/b07ea2cf7d17b4f12855fce8d9baa36d86c803a6/packages/e2e-tests/specs/editor/blocks/navigation.test.js#L1054 and
* https://github.com/WordPress/gutenberg/blob/a92b73ceb9e58dc600fb16f2865b2716ee67cef0/packages/e2e-tests/specs/editor/blocks/gallery.test.js#L106

I'll fix them up tomorrow

File: packages/e2e-test-utils/src/get-list-view-blocks.js
Comment (andrewserong) [reply packages/e2e-test-utils/src/get-list-view-blocks.js]: I see I've missed a couple of tests that also use XPaths expecting a particular sibling of the label. E.g.

* https://github.com/WordPress/gutenberg/blob/b07ea2cf7d17b4f12855fce8d9baa36d86c803a6/packages/e2e-tests/specs/editor/blocks/navigation.test.js#L1054 and
* https://github.com/WordPress/gutenberg/blob/a92b73ceb9e58dc600fb16f2865b2716ee67cef0/packages/e2e-tests/specs/editor/blocks/gallery.test.js#L106

I'll fix them up tomorrow.

Comment (andrewserong) [reply packages/e2e-test-utils/src/get-list-view-blocks.js]: Resolved in https://github.com/WordPress/gutenberg/commit/7344bed5fbc6c469264fea9bd9e7941933a81b00

File: packages/block-editor/src/components/block-title/use-block-display-title.js
Comment (andrewserong) [inline packages/block-editor/src/components/block-title/use-block-display-title.js:33]: I've switched the function signature here (and all usage) over to a props object so that we don't need to pass `undefined` to `maximumLength` in order to use the `context` string. I think it should also help us a bit if in follow- ups we want to pass additional or more complex props (like if we wanted to support `ellipsizeMode` directly in this hook.

**Comment by andrewserong** [comment]
I think this PR is in a good place for review now — the main trade-off is that we're using the `list-view` context to control the particular label requirements of the list view. I've done a little tidying, too, to switch the `useBlockDisplayTitle` hook over to accepting an object of props instead of params, to neaten things up a bit.

I'll just ping @WordPress/gutenberg-design for visibility surrounding the truncation and maximum width, and the general look and feel of truncated headings in the list view. I've updated the PR description with a before and after screenshot to give a better idea of how it sits next to the Lock icon and reusable blocks.

I'll be back on Tuesday so happy to implement any further feedback then. Thanks all! 🙇

**Comment by jameskoster** [comment]
This looks pretty good to me, and we might even consider doing the same thing for other text blocks.

The one thing I would change is the behavior on lock. As the lock icon appears in List View it increases the width of the entire panel to accommodate the lock button which causes the whole UI to shift a bit. We could avoid that shift if the block name width _decreased_ instead. 

What do you think?

**Comment by andrewserong** [comment]
Thanks for taking a look James!

> and we might even consider doing the same thing for other text blocks.

Yes, once this one lands, I'll take a look at the Paragraph block based on Matías' design (https://github.com/WordPress/gutenberg/pull/41855#discussion_r903556163) 👍

> We could avoid that shift if the block name width decreased instead. What do you think?

Yes, that sounds like a good idea to me, so we can keep the whole panel to `350px` without expanding. The thing to watch out for then, is what's the relationship between a row with the Lock icon in it and the following row that doesn't have the icon in it. At the moment (and on trunk), each of the lock icons (and their negative space if no lock icon is present) all line up nicely 🤔. I'll have a play next week once I'm back, and will see how it looks! 🙂

**Comment by jameskoster** [comment]
I'm not entirely sure what you're asking but I think I'd expect to see something like this:

Screenshot 2022-07-01 at 13 34 22

It's worth bearing in mind that in the future it's possible that we'll display the lock button when hovering over unlocked rows.

**Comment by andrewserong** [comment]
> I'm not entirely sure what you're asking but I think I'd expect to see something like this:

That screenshot answered my question, thank you! I'll have a play with implementing that this week 🙂

**Comment by paaljoachim** [comment]
A big yes from me. It is nice to see auto adjusting the Heading name in List View. 

Screen Shot 2022-07-05 at 00 00 32

Adding the heading name/title to List View gives a better overview of the content seen on the post/page. I do expect to see the same thing for the Paragraph block as well. 

Regarding this screenshot.
Screen Shot 2022-07-05 at 00 06 27

The first heading contains an anchor text link.
Second heading contains a lock icon.
The third heading is selected.

We have an anchor name. Then a lock icon for the second heading.
Can we instead of an anchor name use an HTML anchor icon? Currently we use text for anchor and a lock icon for a locked heading. Can we stick to one approach? Having a double truncation feels a bit much.

Here is a quick test. To give a feel for what it can look like. (Not sure if I like it or not, but it does show a better balance of the anchor and lock icons.)
adding anchor icon

**Comment by jameskoster** [comment]
@paaljoachim that's an interesting idea. I like how it opens an opportunity to edit the anchor from list view (in a modal, like the lock status) rather than having to open the Advanced panel of the block inspector.

That said, it seems like something to address in a different issue, what do you think?

**Comment by paaljoachim** [comment]
> That said, it seems like something to address in a different issue, what do you think?

Agreed.

File: packages/block-editor/src/components/list-view/style.scss
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/style.scss]: Reduced this right padding by `1px` so that numerically each of the different widths of the components in the row don't exceed the default `350px` for the panel (without this change, the panel winds up being `351px`). `4px` seems a bit more consistent with other spacing values in components, and also is a bit neater than forcing the max-width of the label wrapper to be `239px`. Happy to move this around if anyone feels strongly about it.

**Comment by andrewserong** [comment]
Thanks for the feedback, folks! @jameskoster I've moved the Lock icon to be within the wrapper, and increased the max-width so there's a little more room without extending the panel's width beyond `350px`. Here's how it's looking now:

image

One potential issue is that it seems to be working nicely with headings at the root level. However, in a nested situation, with a long heading, the new `max-width` does let the List View extend a little ways before being truncated. There's a potential trade-off here — it's probably good to ensure that it doesn't get cut off _too_ short, but I also wasn't sure how comfortable we are with the list view being extended out, so happy for feedback on this, too:



Let me know if there are any further tweaks to be made!

**Comment by jameskoster** [comment]
Thanks @andrewserong, the locking action feels a lot smoother now :) 

I'm not entirely convinced that the increased max-width is necessary though. It would be excellent to definitively address List View behaviour to account for heavy nesting / long names, but it's probably not something we need to handle in this PR. It needs some more design thought and research to get right.

File: packages/block-editor/src/components/list-view/style.scss
Comment (jameskoster) [reply packages/block-editor/src/components/list-view/style.scss]: +1. We can even use `$grid-unit-05`.

**Comment by apeatling** [comment]
> I'm not entirely convinced that the increased max-width is necessary though

I agree with this. I also think you lose more value by reducing the canvas width than you gain by increasing the list view width to accommodate longer titles.

**Comment by andrewserong** [comment]
Thanks again for the feedback, folks, very helpful! I'll take a look and see if I can get the width setting to work correctly for truncating the _available_ width, rather than extending out the list view. Feels like we're getting closer 🙂

File: packages/block-editor/src/components/list-view/style.scss
Comment (andrewserong) [reply packages/block-editor/src/components/list-view/style.scss]: Oh, good call, I'll switch this over to that variable 👍

**Comment by andrewserong** [comment]
I've tweaked the styling a little more in https://github.com/WordPress/gutenberg/pull/41855/commits/dc7d2e349fd7f5da0ab88c6f3e529f58d070b2fe — in order to ensure that the title doesn't unexpectedly extend the panel, I wound up needing to go with an extra wrapper and absolute positioning so that `Truncate` uses the available width without extending the parent. Seems to be working decently now, and the `minWidth` value ensures the title doesn't become unreadable at very deep nesting. Looks like it now won't extend the panel until we get to around 4-5 levels of nesting, which seems a fair bit safer. Here's how it's looking now:



Happy to continue tweaking!

**Comment by jameskoster** [comment]
Thanks for the continued work on this :)

I'm seeing the lock icon extend the panel width again after the latest updates 🙈

https://user-images.githubusercontent.com/846565/177762471-57b69dcb-e07d-46c0-aded-4be599728448.mp4

Edit: The spacing between the lock icon and the ellipsis is unnecessarily large as well. It seems to be caused by the padding on `.block-editor-list-view-leaf .block-editor-list-view-block-contents`.

**Comment by andrewserong** [comment]
Thanks @jameskoster — good catch, I'd put the `min-width` on the wrong container for the lock icon, so I've moved it up to the next parent now, so that it contains the lock icon, too. Toggling the lock setting at deep nesting now no longer affects the panel width:



I've reduced the (effective) right padding on the lock icon by reducing the `.block-editor-list-view-block-contents` padding down to `4px` from `12px`, which feels a bit nicer to my eyes (`4px` still leaves us a little breathing room for the ellipsis characters):

| Icon padding before | Icon padding after |
| --- | --- |
| image | image |

> Thanks for the continued work on this :)

Happy to continue iterating on this, it isn't too much trouble to jump into this PR each day between other tasks, and it's good to try to get it right. I gotta admit, the CSS on this one has been much more fiddly than I'd anticipated! Let me know if there are specific `px` values (or otherwise) that you'd like to go with — I'm mostly tweaking things to a trade-off that feels good enough to me in the editor, so very much appreciate all the feedback here 😊

**Comment by talldan** [comment]
From testing just now - anchor text truncates a little early and doesn't show many characters:

Before:
Screen Shot 2022-07-08 at 2 57 02 pm

After:
Screen Shot 2022-07-08 at 2 57 21 pm

It feels like it should take up any available space. It's probably really tricky to account for all the different situations.

**Comment by andrewserong** [comment]
> It feels like it should take up any available space. It's probably really tricky to account for all the different situations.

It sure is! There's kind of a delicate balancing act between which things should fill up which available space and when 😆
I can have another go at tweaking that one on Monday (it might be that anchor should be in the same wrapper as title, but not the same wrapper as lock  🤔). Thanks for testing!

**Review by talldan — APPROVED** [review]
Works really well! Thanks for sticking with this.

I left an earlier comment about the anchor. If it's an easy fix to improve that before merging, it'd be great. I think it should be possible through some flexboxery.

If it's complicated it might be worth leaving it to a separate PR.

**Review by jameskoster — APPROVED** [review]
Agreed on the anchor fix, otherwise this seems to be working well! Nice job :)

**Comment by andrewserong** [comment]
Thanks for the reviews, folks! It turned out not to be too hard to fix up the anchor, too — I wound up using `flex: 1` on the title wrapper so that it isn't quite as greedy at using up the available space, and then used a `min()` function on the `max-width` of the anchor so that it never uses up more than `40%` of the available space (so that _it_ isn't too greedy with taking up space in narrow titles 😄)

Seems to be working nicely now, so I'll merge this in once tests pass:



Happy to continue tweaking in follow-ups, of course — I'll have a go at the Paragraph block next. Thanks again! 🙇

**Comment by mtias** [comment]
We could consider not rendering the anchor label on headings when they match the same name. I also think it might be useful to add a "Copy anchor" to the ellipsis menu.

**Pull request merged**
